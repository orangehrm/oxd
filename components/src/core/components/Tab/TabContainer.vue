<!--
/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<script lang="ts">
import {h, VNode, KeepAlive, onMounted, Transition, defineComponent} from 'vue';

export default defineComponent({
  name: 'OxdTabContainer',

  props: {
    modelValue: {
      type: String,
      required: false,
      default: null,
    },
    showTabs: {
      type: Boolean,
      required: false,
      default: true,
    },
    animation: {
      type: Boolean,
      required: false,
      default: true,
    },
    keepAlive: {
      type: Boolean,
      required: false,
      default: false,
    },
    cacheAge: {
      type: Number,
      required: false,
      default: 10,
    },
  },

  emits: ['update:modelValue', 'clickTab'],

  setup(props, context) {
    let content: VNode[] = [];

    onMounted(() => {
      if (location.hash) {
        const hash = location.hash.replace(/^#/, '');
        const selectedTab = content.find(
          (tab) => hash === tab.props?.name.replace(/\s+/g, '-').toLowerCase(),
        );
        if (selectedTab) {
          context.emit('update:modelValue', selectedTab.props?.name);
        }
      }
    });

    return () => {
      content = context.slots.default?.() || [];

      const generatePanels = (isKeepAlive: boolean) => {
        const activeTab = content.find((tab, index) => {
          if (!props.modelValue && index === 0) return true;
          return tab.props?.name === props.modelValue;
        });
        return isKeepAlive
          ? h(KeepAlive, {age: props.cacheAge}, activeTab)
          : activeTab;
      };

      return h('div', {class: 'oxd-tab'}, [
        ...(props.showTabs
          ? [
              h(
                'div',
                {class: 'oxd-tab-bar'},
                content.map((tab, index) =>
                  h(
                    'a',
                    {
                      class: {
                        'oxd-tab-segment': true,
                        'oxd-tab-segment--active':
                          props.modelValue === tab.props?.name ||
                          (props.modelValue === null && index === 0),
                        'oxd-tab-segment--disabled': tab.props?.disabled,
                      },
                      href: `#${tab.props?.name
                        .replace(/\s+/g, '-')
                        .toLowerCase()}`,
                      onClick: ($e: MouseEvent) => {
                        if (tab.props?.disabled) {
                          $e.preventDefault();
                        } else {
                          context.emit('clickTab', $e, tab.props?.name);
                          context.emit('update:modelValue', tab.props?.name);
                        }
                      },
                    },
                    tab.props?.name || index,
                  ),
                ),
              ),
            ]
          : []),
        ...(props.animation
          ? [
              h(
                Transition,
                {
                  name: 'transition-slide-right',
                },
                {
                  default: () => generatePanels(props.keepAlive),
                },
              ),
            ]
          : [generatePanels(props.keepAlive)]),
      ]);
    };
  },
});
</script>

<style src="./tab-container.scss" lang="scss" scoped></style>
