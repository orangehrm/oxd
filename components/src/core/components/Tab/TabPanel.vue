<!--
/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<script lang="ts">
import {
  h,
  onActivated,
  onDeactivated,
  onBeforeMount,
  onBeforeUnmount,
  defineComponent,
} from 'vue';

export default defineComponent({
  name: 'oxd-tab-panel',
  props: {
    name: {
      type: String,
      required: true,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  emits: ['activate', 'deactivate'],
  setup(props, context) {
    onActivated(() => context.emit('activate'));
    onBeforeMount(() => context.emit('activate'));
    onDeactivated(() => context.emit('deactivate'));
    onBeforeUnmount(() => context.emit('deactivate'));

    return () =>
      h(
        'div',
        {
          id: props.name,
          class: 'oxd-tab-panel',
        },
        context.slots.default?.(),
      );
  },
});
</script>
