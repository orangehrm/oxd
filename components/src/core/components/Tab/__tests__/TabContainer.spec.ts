/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
import {h} from 'vue';
import {mount} from '@vue/test-utils';
import TabPanel from '@ohrm/oxd/core/components/Tab/TabPanel.vue';
import TabContainer from '@ohrm/oxd/core/components/Tab/TabContainer.vue';

const Tabs = [
  h(
    TabPanel,
    {name: 'tab1'},
    {
      default: () => h('div', 'Tab 1'),
    },
  ),
  h(
    TabPanel,
    {name: 'tab2'},
    {
      default: () => h('div', 'Tab 2'),
    },
  ),
  h(
    TabPanel,
    {name: 'tab3'},
    {
      default: () => h('div', 'Tab 3'),
    },
  ),
];

describe('TabContainer.vue', () => {
  it('renders OXD Tab Container', () => {
    const wrapper = mount(TabContainer, {
      slots: {
        default: () => Tabs,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render OXD Tab Container with tabs', async () => {
    const wrapper = mount(TabContainer, {
      slots: {
        default: () => Tabs,
      },
    });
    const tabs = wrapper.findAll('.oxd-tab-segment');
    expect(tabs.length).toBe(3);
    expect(tabs[0].classes()).toContain('oxd-tab-segment--active');
  });

  it('should change current tab on click', async () => {
    const wrapper = mount(TabContainer, {
      slots: {
        default: () => Tabs,
      },
    });
    const tabs = wrapper.findAll('.oxd-tab-segment');
    await tabs[2].trigger('click');
    expect(wrapper.emitted('update:modelValue')).toEqual([['tab3']]);
  });

  it('should not change current tab on click disabled tab', async () => {
    const wrapper = mount(TabContainer, {
      slots: {
        default: () => [
          h(
            TabPanel,
            {name: 'tab1'},
            {
              default: () => h('div', 'Tab 1'),
            },
          ),
          h(
            TabPanel,
            {name: 'tab2', disabled: true},
            {
              default: () => h('div', 'Tab 2'),
            },
          ),
        ],
      },
    });
    const tabs = wrapper.findAll('.oxd-tab-segment');
    await tabs[1].trigger('click');
    expect(wrapper.emitted('update:modelValue')).toBeFalsy();
  });

  it('should select current tab by default', async () => {
    const wrapper = mount(TabContainer, {
      props: {
        modelValue: 'tab2',
      },
      slots: {
        default: () => Tabs,
      },
    });
    const tabs = wrapper.findAll('.oxd-tab-segment');
    expect(tabs[1].classes()).toContain('oxd-tab-segment--active');
  });

  it('should select current tab by url', async () => {
    // eslint-disable-next-line no-global-assign
    window = Object.create(window);
    Object.defineProperty(window, 'location', {
      value: {
        href: 'http://localhost#tab2',
        hash: '#tab2',
      },
    });

    const wrapper = mount(TabContainer, {
      slots: {
        default: () => Tabs,
      },
    });
    expect(wrapper.emitted('update:modelValue')).toEqual([['tab2']]);
  });
});
