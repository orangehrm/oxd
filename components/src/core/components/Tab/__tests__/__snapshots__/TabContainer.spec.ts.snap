// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabContainer.vue renders OXD Tab Container 1`] = `
<div class="oxd-tab">
  <div class="oxd-tab-bar"><a class="oxd-tab-segment oxd-tab-segment--active" href="#tab1">tab1</a><a class="oxd-tab-segment" href="#tab2">tab2</a><a class="oxd-tab-segment" href="#tab3">tab3</a></div>
  <transition-stub>
    <div id="tab1" class="oxd-tab-panel">
      <div>Tab 1</div>
    </div>
  </transition-stub>
</div>
`;
